### PR TITLE
builtin_macros: Migrate diagnostic

### PR DESCRIPTION
**Pull request by wonchulee**
* Migrate the `rustc_biltin_macros` crate's diagnostic to translatable diagnostic structs.

### CODE REVIEW

**Comment by Nilstrieb** [comment]
Hi, I've seen you changed some diagnostic structs in your PR. After #103345, the way we refer to fluent messages changed. They are now in a flat namespace with the same identifier as in the fluent file. For example,  `parser::cool_thing` is now `parser_cool_thing` and `parser::suggestion` just `suggestion`.
You should [rebase](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts) to the latest master and change your fluent message references as described above. Thanks!

File: compiler/rustc_builtin_macros/src/errors.rs [hunk 1]
```diff
@@ -0,0 +1,620 @@
...
+}
+
+#[derive(SessionDiagnostic)]
+#[diag(builtin_macros::asm_duplicate_option)]
+pub(crate) struct AsmDuplicateOption {
+    pub(crate) symbol: Symbol,
+    #[primary_span]
+    #[label]
+    pub(crate) span: Span,
+    // FIXME: This should be a `tool_only_span_suggestion`
```
Comment (davidtwco) [inline compiler/rustc_builtin_macros/src/errors.rs:101]: I think we've got support for this now after #103575.

File: compiler/rustc_builtin_macros/src/errors.rs [hunk 2]
```diff
@@ -0,0 +1,620 @@
...
+    pub span: Span,
+}
+
+pub enum Snippet {
+    ByteCharacter { span: Span, snippet: String },
+    ByteString { span: Span, snippet: String },
+    WrappingNumberInArray { span: Span, snippet: String },
+}
+
+impl AddSubdiagnostic for Snippet {
```
Comment (davidtwco) [inline compiler/rustc_builtin_macros/src/errors.rs:308]: I think that you can use the derive for this.

**Review by davidtwco — COMMENTED** [review]
Apologies for the delay in reviewing this, this is a great start - I assume from the CI failures that there are some diagnostics still to migrate.

**Comment by Dylan-DPC** [comment]
@wonchulee any updates on this?

**Comment by wonchulee** [comment]
> @wonchulee any updates on this?

Hello, This is still a work in progress, and as I mentioned in https://github.com/rust-lang/rust/issues/100717, I don't have enough time to finish it these days. If someone would like to take it on, that would be great. I'll try to keep going as long as I can until then.

**Comment by jyn514** [comment]
Hi @wonchulee, I'm going to close this since you're not sure when you'll have time to work on it. Thank you for the effort you've put in already, and feel free to pick up the work later if you have more time :) (although do be aware that https://github.com/rust-lang/rust/pull/110092 migrated a lot of the diagnostics in the meantime, so it might make sense to pick a crate other than rustc_builtin_macros).

**Pull request closed without merging**